pkg/resources/openstack/lb.go

```diff
@@ -17,43 +17,93 @@ limitations under the License.
 package openstack
 
 import (
-	"fmt"
+	"strings"
 
+	"github.com/gophercloud/gophercloud/openstack/loadbalancer/v2/listeners"
 	"github.com/gophercloud/gophercloud/openstack/loadbalancer/v2/loadbalancers"
+	v2pools "github.com/gophercloud/gophercloud/openstack/loadbalancer/v2/pools"
 	"k8s.io/kops/pkg/resources"
 	"k8s.io/kops/upup/pkg/fi"
 	"k8s.io/kops/upup/pkg/fi/cloudup/openstack"
 )
 
 const (
-	typeLB = "lb"
+	typeLB  = "LoadBalancer"
+	typeLBL = "LBListener"
+	typeLBP = "LBPool"
 )
 
-func listLBs(cloud openstack.OpenstackCloud, clusterName string) ([]*resources.Resource, error) {
+func (os *clusterDiscoveryOS) ListLB() ([]*resources.Resource, error) {
+	var resourceTrackers []*resources.Resource
 	opts := loadbalancers.ListOpts{
-		Name: clusterName,
+		Name: "api." + os.clusterName,
 	}
-	lbs, err := cloud.ListLBs(opts)
+	lbs, err := os.osCloud.ListLBs(opts)
 	if err != nil {
-		return nil, fmt.Errorf("failed to list lbs: %s", err)
+		return nil, err
 	}
 
-	var rts []*resources.Resource
-	for _, t := range lbs {
-		rt := &resources.Resource{
-			Name: t.Name,
-			ID:   t.ID,
+	for _, lb := range lbs {
+		resourceTracker := &resources.Resource{
+			Name: lb.Name,
+			ID:   lb.ID,
 			Type: typeLB,
 			Deleter: func(cloud fi.Cloud, r *resources.Resource) error {
 				opts := loadbalancers.DeleteOpts{
 					Cascade: true,
 				}
-				return loadbalancers.Delete(cloud.(openstack.OpenstackCloud).LoadBalancerClient(), t.ID, opts).ExtractErr()
+				return cloud.(openstack.OpenstackCloud).DeleteLB(r.ID, opts)
 			},
-			Obj: lbs,
 		}
-		rts = append(rts, rt)
+		resourceTrackers = append(resourceTrackers, resourceTracker)
+	}
+	return resourceTrackers, nil
+}
+
+func (os *clusterDiscoveryOS) ListLBPools() ([]*resources.Resource, error) {
+	var resourceTrackers []*resources.Resource
+
+	pools, err := os.osCloud.ListPools(v2pools.ListOpts{})
+	if err != nil {
+		return nil, err
 	}
 
-	return rts, nil
+	for _, pool := range pools {
+		if strings.Contains(pool.Name, os.clusterName) {
+			resourceTracker := &resources.Resource{
+				Name: pool.Name,
+				ID:   pool.ID,
+				Type: typeLBP,
+				Deleter: func(cloud fi.Cloud, r *resources.Resource) error {
+					return cloud.(openstack.OpenstackCloud).DeletePool(r.ID)
+				},
+			}
+			resourceTrackers = append(resourceTrackers, resourceTracker)
+		}
+	}
+	return resourceTrackers, nil
+}
+
+func (os *clusterDiscoveryOS) ListLBListener() ([]*resources.Resource, error) {
+	var resourceTrackers []*resources.Resource
+
+	listeners, err := os.osCloud.ListListeners(listeners.ListOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, listener := range listeners {
+		if strings.Contains(listener.Name, os.clusterName) {
```

Comment:
Would be nice to have something stronger than a substring match in future - feels very likely to false-positive...

But I think OpenStack has projects to isolate multiple clusters, so maybe it doesn't matter at much?

Reply:
at least our plan is to use one kops cluster per one openstack project. So it does not really match. However, if people is planning to deploy 100 clusters to same project, the naming should be planned. Otherwise this could make some problems